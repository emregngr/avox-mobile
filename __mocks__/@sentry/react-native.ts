import type { ComponentType } from 'react'

const mockedSentry = {
  init: jest.fn(),
  wrap: <T extends ComponentType<any>>(Component: T): T => Component,
  withScope: jest.fn((callback: (scope: any) => void) =>
    callback({
      setTag: jest.fn(),
      setLevel: jest.fn(),
      setUser: jest.fn(),
      setContext: jest.fn(),
    }),
  ),
  captureException: jest.fn(),
  captureMessage: jest.fn(),
  addBreadcrumb: jest.fn(),
  configureScope: jest.fn(),
  setUser: jest.fn(),
  setTag: jest.fn(),
  setContext: jest.fn(),
  setRelease: jest.fn(),
  setDist: jest.fn(),
  setEnvironment: jest.fn(),
  close: jest.fn(() => Promise.resolve(true)),
  flush: jest.fn(() => Promise.resolve(true)),
}

module.exports = mockedSentry

export type mockedSentryType = typeof mockedSentry
