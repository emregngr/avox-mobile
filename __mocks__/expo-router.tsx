import { createElement, Fragment, ReactNode } from 'react'
import { View } from 'react-native'

const mockedRouterFunctions = {
  push: jest.fn(),
  navigate: jest.fn(),
  replace: jest.fn(),
  back: jest.fn(),
  canGoBack: jest.fn(() => true),
  setParams: jest.fn(),
}

let mockedPathname = '/home'

const setMockPathname = (pathname: string) => {
  mockedPathname = pathname
  const usePathnameMock = module.exports.usePathname
  if (usePathnameMock && usePathnameMock.mockReturnValue) {
    usePathnameMock.mockReturnValue(mockedPathname)
  }
}

const clearAllMocks = () => {
  Object.values(mockedRouterFunctions).forEach(fn => {
    if (typeof fn === 'function' && fn.mockClear) {
      fn.mockClear()
    }
  })
}

const MockScreen = ({ children }: { children: ReactNode }) =>
  createElement(Fragment, null, children)

const MockedStack = ({ children }: { children: ReactNode }) =>
  createElement(Fragment, null, children)

const MockedTabs = ({
  children,
  screenOptions,
  ...props
}: {
  children: ReactNode
  screenOptions?: any
  [key: string]: any
}) => (
  <View testID="tabs" data-screen-options={JSON.stringify(screenOptions)} {...props}>
    {children}
  </View>
)

const MockTabScreen = ({ name, ...props }: { name: string; [key: string]: any }) => (
  <View testID={`tab-screen-${name}`} data-name={name} {...props} />
)

module.exports = {
  router: mockedRouterFunctions,
  useRouter: jest.fn(() => mockedRouterFunctions),
  useFocusEffect: jest.fn(),
  useSegments: jest.fn(() => []),
  useSearchParams: jest.fn(() => ({})),
  useLocalSearchParams: jest.fn(() => ({})),
  useGlobalSearchParams: jest.fn(() => ({})),
  usePathname: jest.fn(() => mockedPathname),
  useNavigation: jest.fn(() => ({})),
  Slot: ({ children }: { children: ReactNode }) => <>{children}</>,
  Stack: Object.assign(MockedStack, { Screen: MockScreen }),
  Tabs: Object.assign(MockedTabs, { Screen: MockTabScreen }),
  Link: ({ children, ...props }: { children: ReactNode }) =>
    createElement(Fragment, null, children),
  Redirect: () => null,
  setMockPathname: setMockPathname,
  clearAllMocks: clearAllMocks,
  mockedFunctions: mockedRouterFunctions,
  get mockPathname() {
    return mockedPathname
  },
}
