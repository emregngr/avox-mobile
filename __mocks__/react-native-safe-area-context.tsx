import { ComponentType, ReactNode } from 'react'
import { View } from 'react-native'

export type Insets = {
  top: number
  bottom: number
  left: number
  right: number
}

export type Frame = {
  x: number
  y: number
  width: number
  height: number
}

type SafeAreaConsumerProps = {
  children: (insets: Insets) => ReactNode
}

export const SafeAreaProvider = ({ children }: { children: ReactNode }) => <>{children}</>

export const SafeAreaConsumer = ({ children }: SafeAreaConsumerProps) =>
  children({ top: 0, bottom: 0, left: 0, right: 0 })

export const SafeAreaView = View

export const useSafeAreaInsets = (): Insets => ({
  top: 0,
  bottom: 0,
  left: 0,
  right: 0,
})

export const mockedUseSafeAreaInsets = jest.fn().mockReturnValue({
  top: 0,
  bottom: 0,
  left: 0,
  right: 0,
})

export const useSafeAreaFrame = (): Frame => ({
  x: 0,
  y: 0,
  width: 375,
  height: 812,
})

export function withSafeAreaInsets<P extends object>(
  Component: ComponentType<P>,
): ComponentType<P> {
  return Component
}
