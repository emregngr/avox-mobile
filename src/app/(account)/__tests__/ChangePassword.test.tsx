import { fireEvent, render } from '@testing-library/react-native'
import { router } from 'expo-router'
import type { ReactNode } from 'react'
import React from 'react'
import { SafeAreaProvider } from 'react-native-safe-area-context'

import PasswordScreen from '@/app/(account)/password'

const { mockedAuth } = require('@react-native-firebase/auth')

jest.mock('@/locales/i18next', () => ({
  getLocale: (key: string) => key,
}))

jest.mock('@/components/common', () => {
  const { View, Text, TouchableOpacity } = require('react-native')

  return {
    SafeLayout: ({ children }: { children: ReactNode }) => (
      <View testID="safe-layout">{children}</View>
    ),

    Header: ({ title, backIconOnPress }: { title: string; backIconOnPress: () => void }) => (
      <>
        <Text testID="header-title">{title}</Text>
        <TouchableOpacity onPress={backIconOnPress} testID="back-button">
          <Text>Back</Text>
        </TouchableOpacity>
      </>
    ),
  }
})

jest.mock('@/components/feature', () => {
  const { Text } = require('react-native')

  return {
    AddPassword: () => <Text>AddPassword Component</Text>,

    ChangePassword: () => <Text>ChangePassword Component</Text>,
  }
})

const renderWithSafeAreaProvider = (component: ReactNode) =>
  render(
    <SafeAreaProvider
      initialMetrics={{
        insets: { top: 0, left: 0, right: 0, bottom: 0 },
        frame: { x: 0, y: 0, width: 0, height: 0 },
      }}
    >
      {component}
    </SafeAreaProvider>,
  )

describe('Password Screen', () => {
  it('should render ChangePassword component for a password user', () => {
    mockedAuth.currentUser = {
      providerData: [{ providerId: 'password' }],
    }

    const { getByText, queryByText } = renderWithSafeAreaProvider(<PasswordScreen />)

    expect(getByText('changePassword')).toBeTruthy()
    expect(getByText('ChangePassword Component')).toBeTruthy()
    expect(queryByText('AddPassword Component')).toBeNull()
  })

  it('should render AddPassword component for a non-password user', () => {
    mockedAuth.currentUser = {
      providerData: [{ providerId: 'google.com' }],
    }

    const { getByText, queryByText } = renderWithSafeAreaProvider(<PasswordScreen />)

    expect(getByText('addPassword')).toBeTruthy()
    expect(getByText('AddPassword Component')).toBeTruthy()
    expect(queryByText('ChangePassword Component')).toBeNull()
  })

  it('should call router.back when the header back button is pressed', () => {
    mockedAuth.currentUser = {
      providerData: [{ providerId: 'password' }],
    }

    const { getByTestId } = renderWithSafeAreaProvider(<PasswordScreen />)

    const backButton = getByTestId('back-button')
    fireEvent.press(backButton)

    expect(router.back).toHaveBeenCalledTimes(1)
  })
})

describe('Password Screen Snapshot', () => {
  it('should render the Password Screen successfully', () => {
    const { toJSON } = renderWithSafeAreaProvider(<PasswordScreen />)

    expect(toJSON()).toMatchSnapshot()
  })
})
