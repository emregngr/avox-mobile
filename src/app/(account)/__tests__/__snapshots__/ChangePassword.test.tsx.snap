// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Password Screen Snapshot should render the Password Screen successfully 1`] = `
<View
  testID="safe-layout"
>
  <Text
    testID="header-title"
  >
    changePassword
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="back-button"
  >
    <Text>
      Back
    </Text>
  </View>
  <Text>
    ChangePassword Component
  </Text>
</View>
`;
